### PR TITLE
Leader lane multi-chain gas price reports

### PR DESCRIPTION
**Pull request by matYang**
Previously, leader lane ORM was added to enabled Commit plugin to read multi-chain gas prices during `Observation`. This PR updates `Report` phase so Commit plugin can now report multi-chain gas prices.

Important changes are:
1. make use of the PriceReportingDisabled flag to skip price reports on non-leader lanes
2. update gas price logic during Report to handle multi-chain gas prices

### CODE REVIEW

File: .changeset/sour-elephants-sell.md
Comment (makramkd) [inline .changeset/sour-elephants-sell.md:1]: We're not committing any changesets anymore

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 1]
```diff
@@ -112,15 +112,22 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
 	}
 
-	sourceGasPriceUSD, tokenPricesUSD, err := r.observePriceUpdates(ctx)
+	// Fetches multi-lane gasPricesUSD and tokenPricesUSD for the same dest chain
+	gasPricesUSD, tokenPricesUSD, err := r.observePriceUpdates(ctx)
 	if err != nil {
 		return nil, err
 	}
 
+	// For backwards compatibility with the order release during phased rollout, include the default gas price on this lane
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]:
```suggestion
	// For backwards compatibility with the older release during phased rollout, include the default gas price on this lane
```

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 2]
```diff
@@ -134,11 +141,29 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
+		SourceGasPriceUSD:         defaultGasPricesUSD,
+		SourceGasPriceUSDPerChain: gasPricesUSD,
 	}.Marshal()
 }
 
+func (r *CommitReportingPlugin) observePriceUpdates(
+	ctx context.Context,
+) (gasPricesUSD map[uint64]*big.Int, tokenPricesUSD map[cciptypes.Address]*big.Int, err error) {
+	// Do not observe prices if price reporting is disabled. Price reporting will be disabled for lanes that are not leader lanes.
+	if r.offchainConfig.PriceReportingDisabled {
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:150]: Is it worth logging in these scenarios or nah?

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:150]: Also sidenote I wish this was `PriceReportingEnabled` instead

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 3]
```diff
@@ -384,29 +401,38 @@ func extractObservationData(lggr logger.Logger, f int, observations []ccip.Commi
...
-	// Observations are invalid if observed gas price is nil, we require at least f+1 valid observations.
-	if len(gasPrices) <= f {
-		return nil, nil, nil, fmt.Errorf("not enough valid observations with non-nil gas prices: #obs=%d, f=%d", len(gasPrices), f)
+	gasPrices = make(map[uint64][]*big.Int)
+	for selector, perChainPriceObservations := range gasPriceObservations {
+		// Price is dropped if there are not enough valid observations. With a threshold of 2*(f-1) + 1, we achieve a balance between safety and liveness.
+		// During phased-rollout where some honest nodes may not have started observing the token yet, it requires 5 malicious node with 1 being the leader to successfully alter price.
+		// During regular operation, it requires 3 malicious nodes with 1 being the leader to temporarily delay price update for the token.
+		if len(perChainPriceObservations) < (2*(f-1) + 1) {
+			lggr.Warnf("Skipping chain with selector %d due to not enough valid observations: #obs=%d, f=%d", selector, len(perChainPriceObservations), f)
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: Worth logging the actual `2f-1` value?

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 4]
```diff
@@ -449,30 +476,38 @@ func (r *CommitReportingPlugin) calculatePriceUpdates(gasPriceObs []*big.Int, to
...
+		// Default to updating so that we update if there are no prior updates.
+		latestGasPrice, exists := latestGasPrice[chainSelector]
+		if exists && latestGasPrice.value != nil {
+			gasPriceUpdatedRecently := time.Since(latestGasPrice.timestamp) < r.offchainConfig.GasPriceHeartBeat
+			gasPriceDeviated, err := r.gasPriceEstimator.Deviates(newGasPrice, latestGasPrice.value)
+			if err != nil {
+				return nil, nil, err
+			}
+			if gasPriceUpdatedRecently && !gasPriceDeviated {
+				r.lggr.Debugw("gas price was updated recently, skipping the update",
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: To reflect the condition, maybe add `and not deviated sufficiently`?

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 5]
```diff
@@ -449,30 +476,38 @@ func (r *CommitReportingPlugin) calculatePriceUpdates(gasPriceObs []*big.Int, to
...
 	var gasPriceUpdate []cciptypes.GasPrice
-	// Default to updating so that we update if there are no prior updates.
-	shouldUpdate := true
-	if latestGasPrice.value != nil {
-		gasPriceUpdatedRecently := time.Since(latestGasPrice.timestamp) < r.offchainConfig.GasPriceHeartBeat
-		gasPriceDeviated, err := r.gasPriceEstimator.Deviates(newGasPrice, latestGasPrice.value)
+	for chainSelector, gasPriceObservations := range gasPriceObs {
+		newGasPrice, err := r.gasPriceEstimator.Median(gasPriceObservations) // Compute the median price
 		if err != nil {
 			return nil, nil, err
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: Do these returned `err` values have enough context to figure out where we failed exactly? This has happened before where its hard to deduce the exact place things stopped working

File: core/services/ocr2/plugins/ccip/internal/ccipdb/price_service.go [hunk 6]
```diff
@@ -38,6 +38,7 @@ type PriceService interface {
 
 	// GetGasAndTokenPrices fetches source chain gas prices and relevant token prices from all lanes that touch the given dest chain.
 	// The prices have been written into the DB by each lane's PriceService in the background. The prices are denoted in USD.
+	// If the results are empty, they are in the form of empty maps, as opposed to being nil.
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/internal/ccipdb/price_service.go]: This is an interesting API contract

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 7]
```diff
@@ -134,11 +141,29 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
+func (r *CommitReportingPlugin) observePriceUpdates(
+	ctx context.Context,
+) (gasPricesUSD map[uint64]*big.Int, tokenPricesUSD map[cciptypes.Address]*big.Int, err error) {
+	// Do not observe prices if price reporting is disabled. Price reporting will be disabled for lanes that are not leader lanes.
+	if r.offchainConfig.PriceReportingDisabled {
+		return map[uint64]*big.Int{}, map[cciptypes.Address]*big.Int{}, nil
+	}
+
+	// Fetches multi-lane gas prices and token prices, for the given dest chain
+	gasPricesUSD, tokenPricesUSD, err = r.priceService.GetGasAndTokenPrices(ctx, r.destChainSelector)
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:156]: Worth checking that `gasPricesUSD` and `tokenPricesUSD` are not nil and erroring if they are? Thats a breakage in the API contract

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go (hunk 2 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:150]: yeah me too, this was defined in April under a different leader lane architecture that got killed due to LOOPPS, but Engops already update the tooling to include this flag, didn't wanna change it again

File: core/services/ocr2/plugins/ccip/internal/ccipdb/price_service.go (hunk 6 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/internal/ccipdb/price_service.go]: cleaned up, added nil checks on ocr2.go

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go (hunk 7 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:156]: after the API cleanup, added explicit nil check

**Review by makramkd — APPROVED** [review]

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 8]
```diff
@@ -112,15 +112,25 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
 
-	sourceGasPriceUSD, tokenPricesUSD, err := r.observePriceUpdates(ctx)
+	// Fetches multi-lane gasPricesUSD and tokenPricesUSD for the same dest chain
+	gasPricesUSD, tokenPricesUSD, err := r.observePriceUpdates(ctx)
 	if err != nil {
 		return nil, err
 	}
+	// Set prices to empty maps if nil to be friendlier to JSON encoding
+	if gasPricesUSD == nil {
+		gasPricesUSD = map[uint64]*big.Int{}
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: maybe better to do this inside `r.observePriceUpdates(ctx)`

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 9]
```diff
@@ -112,15 +112,25 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
+	}
+	if tokenPricesUSD == nil {
+		tokenPricesUSD = map[cciptypes.Address]*big.Int{}
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: same

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 10]
```diff
@@ -112,15 +112,25 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
+	}
+	// For backwards compatibility with the older release during phased rollout, set the default gas price on this lane
+	defaultGasPricesUSD := gasPricesUSD[r.sourceChainSelector]
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: rename var to `defaultGasPriceUSD` and check if it exists in the map?

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 11]
```diff
@@ -134,11 +144,32 @@ func (r *CommitReportingPlugin) Observation(ctx context.Context, epochAndRound t
...
+	// Do not observe prices if price reporting is disabled. Price reporting will be disabled for lanes that are not leader lanes.
+	if r.offchainConfig.PriceReportingDisabled {
+		r.lggr.Infow("Price reporting disabled, skipping gas and token price reads")
+		return nil, nil, nil
+	}
+
+	// Fetches multi-lane gas prices and token prices, for the given dest chain
+	gasPricesUSD, tokenPricesUSD, err = r.priceService.GetGasAndTokenPrices(ctx, r.destChainSelector)
+	if err != nil {
+		return nil, nil, err
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: wrap err?

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 12]
```diff
@@ -219,33 +233,34 @@ func (r *CommitReportingPlugin) getLatestTokenPriceUpdates(ctx context.Context,
...
-	for _, priceUpdate := range gasPriceUpdates {
+	latestUpdates := make(map[uint64]update)
+	for _, gasUpdate := range gasPriceUpdates {
+		priceUpdate := gasUpdate.GasPriceUpdate
 		// Ordered by ascending timestamps
-		timestamp := time.Unix(priceUpdate.GasPriceUpdate.TimestampUnixSec.Int64(), 0)
-		if !timestamp.Before(gasUpdate.timestamp) {
-			gasUpdate = update{
+		timestamp := time.Unix(priceUpdate.TimestampUnixSec.Int64(), 0)
+		if priceUpdate.Value != nil && !timestamp.Before(latestUpdates[priceUpdate.DestChainSelector].timestamp) {
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:256]: check if `latestUpdates[priceUpdate.DestChainSelector]` exists?

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go (hunk 10 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: renamed to `sourceGasPriceUSD`, it's ok if it does not exist, nil for this value is supported

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go (hunk 12 above)
Comment (matYang) [reply core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go:256]: `update` is a struct, I think it's intended that `latestUpdates[priceUpdate.DestChainSelector]` may not exist, if it doesn't, then timestamp is 0 so `!timestamp.Before` is always true, and that initializes the value

**Review by dimkouv — APPROVED** [review]

**Pull request merged**